support generating separate blob for prefetched files

daiyongxuan opened the pull request:
## Relevant Issue #1589
The objective is to create an independent image layer to prioritize the fetching of necessary blobs during startup, enhancing performance.

## Details
This introduces the "optimize" subcommand, which generates a new blob containing prefetch files and creates a new RAFS format bootstrap. This ensures compatibility with the nydus-image check and optimizes the startup process.

imeoer commented on storage/src/backend/localfs.rs:
```diff
@@ -163,6 +163,7 @@ impl LocalFs {
         }
 
         let blob_file_path = self.get_blob_path(blob_id)?;
+        println!("blob file path: {}", blob_file_path.display());
```
remove useless codes.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1734,6 +1819,32 @@ impl Command {
+        match matches.get_one::<String>("prefetch-files") {
+            Some(v) => {
+                let content = std::fs::read_to_string(v)
+                    .map_err(|_| anyhow!("failed to read prefetch files from {}", v))?;
```
Do not ignore the err.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1634,6 +1673,52 @@ impl Command {
+            blobs_dir_path.to_path_buf(),
+            prefetch_nodes,
+        )
+        .unwrap();
```
Error should be handled instead of calling `unwrap`.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -529,6 +534,36 @@ fn prepare_cmd_args(bti_string: &'static str) -> App {
+            )
+            .arg(arg_config.clone())
+            .arg(
+                Arg::new("blob-dir")
```
Should use `--backend-type`, `--backend-config`, `--backend-config-file` like `compact` subcommand.

imeoer commented on output/chunkdict_bootstrap:
Remove useless files.

daiyongxuan committed:
builder: introduce optimize subcommand for prefetch

We can statically analyze the image entrypoint dependency, or use runtime dynamic
analysis technologies such as ebpf, fanotify, metric, etc. to obtain the container
file access pattern, and then build this part of data into an independent image layer:

* preferentially fetch blob during the image startup phase to reduce network and disk IO.
* avoid frequent image builds, allows for better local cache utilization.

Implement optimize subcommand to optimize image bootstrap
from a prefetch file list, generate a new blob.

```
nydus-image optimize --prefetch-files /path/to/prefetch-files.txt \
  --bootstrap /path/to/bootstrap \
  --blob-dir /path/to/blobs
```
This will generate a new bootstrap and new blob in `blob-dir`.

Signed-off-by: daiyongxuan <daiyongxuan20@mails.ucas.ac.cn>

imeoer commented on smoke/tests/native_layer_test.go:
```diff
@@ -81,6 +81,7 @@ func (n *NativeLayerTestSuite) TestMakeLayers() test.Generator {
 			ctx.Runtime.RafsMode = scenario.GetString(paramRafsMode)
 			ctx.Runtime.EnablePrefetch = scenario.GetBool(paramEnablePrefetch)
 			ctx.Runtime.AmplifyIO = scenario.GetUInt64(paramAmplifyIO)
+			ctx.Runtime.ChunkDedupDb = scenario.GetString(paramChunkDedupDb)
```
This seems like an unnecessary merge commit.

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -53,9 +69,91 @@ pub struct ChunkdictBlobInfo {
+    fn new(ctx: &BuildContext, blob_layer_num: u32, blobs_dir_path: &Path) -> Result<Self> {
+        let mut blob_info = BlobInfo::new(
+            blob_layer_num,
+            String::from("Prefetch-blob"),
```
`prefetch-blob`.

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -53,9 +69,91 @@ pub struct ChunkdictBlobInfo {
+                | BlobFeatures::INLINED_CHUNK_DIGEST
+                | BlobFeatures::HAS_TAR_HEADER
+                | BlobFeatures::HAS_TOC
+                | BlobFeatures::CAP_TAR_TOC,
```
These features should inherit from the base bootstrap's last blob.

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -53,9 +69,91 @@ pub struct ChunkdictBlobInfo {
+                | BlobFeatures::HAS_TOC
+                | BlobFeatures::CAP_TAR_TOC,
+        );
+        blob_info.set_compressor(Algorithm::Zstd);
```
The compressor should inherit from the base bootstrap's last blob.

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -53,9 +69,91 @@ pub struct ChunkdictBlobInfo {
+        );
+        blob_info.set_compressor(Algorithm::Zstd);
+        let mut blob_ctx = BlobContext::from(ctx, &blob_info, ChunkSource::Build)?;
+        blob_ctx.chunk_count = 0;
```
unnecessary set?

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -53,9 +69,91 @@ pub struct ChunkdictBlobInfo {
+        let mut blob_ctx = BlobContext::from(ctx, &blob_info, ChunkSource::Build)?;
+        blob_ctx.chunk_count = 0;
+        blob_ctx.blob_meta_info_enabled = true;
+        let blob_writer = ArtifactWriter::new(crate::ArtifactStorage::SingleFile(
```
Should use `ArtifactStorage::FileDir` here.

imeoer commented on src/bin/nydus-image/main.rs:
```diff
@@ -1734,6 +1819,32 @@ impl Command {
+                }
+                Ok(prefetch_files)
+            }
+            None => panic!("missing parameter `prefetch-files`"),
```
`bail!`.

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -90,6 +188,211 @@ impl Generator {
+        tree: &mut Tree,
+        ctx: &mut BuildContext,
+        bootstrap_mgr: &mut BootstrapManager,
+        blobtable: &mut RafsV6BlobTable,
```
`blob_table`, and need also to be compatible with RAFS v5.

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -90,6 +188,211 @@ impl Generator {
+
+        let mut blob_state =
+            PrefetchBlobState::new(&ctx, blob_layer_num as u32, &blobs_dir_path).unwrap();
+        let mut batch = BatchContextGenerator::new(4096).unwrap();
```
The default batch size should be `0`?

imeoer commented on builder/src/chunkdict_generator.rs:
```diff
@@ -90,6 +188,211 @@ impl Generator {
+        if let Some((_, blob_ctx)) = blob_mgr.get_current_blob() {
+            Blob::dump_meta_data(&ctx, blob_ctx, blob_state.blob_writer.as_mut()).unwrap();
+        } else {
+            panic!();
```
Should return an specified error.

imeoer reviewed (approved):
Thanks, LGTM!
